### PR TITLE
Add brokers to steward

### PR DESCRIPTION
**Pull request by metalmatze**
This is still WIP, just like the rest of the project.
The Router will simply answer with _hey_ to any forwarded Event.

### CODE REVIEW

File: event.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package main
+
+// Event is passed back and forth from and to brokers
+type Event struct {
+	ID      string
+	Body    string
+	User    string
+	Channel string
+	Broker  Broker
```
Comment (fortytw2) [inline event.go]: Are events tied to Brokers? Only in one direction, right?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+// Config gets all values loaded into and is used to configure other structs
+type Config struct {
+	SlackToken string `env:"key=SLACK_TOKEN" arg:"--slack-token"`
+}
+
+func main() {
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+
+	c := Config{}
```
Comment (fortytw2) [inline main.go]: var c Config

File: main.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	// blocks and waits for incoming events on a channel
+	router.Run()
+}
+
+// Broker is anything that can interact with the Router from the outside,
+// e.g. Slack, Telegram, CLI...
+type Broker interface {
+	Name() string
+	Run(in chan<- *Event)
```
Comment (fortytw2) [inline main.go]: Maybe have a 

```go
CheckTimeout(t time.Duration) *Event, error
Send(*Event) error
```
style API. Channels can become difficult to consume and test rather quickly

File: main.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	Send(*Event)
+}
+
+// Router knows about all Brokers and passes Events to the correct handlers
+type Router struct {
+	sync.Mutex
```
Comment (fortytw2) [inline main.go]: `r.Mutex` is publicly exposed, perhaps `brokerMu sync.RWMutex`

File: main.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	r.Brokers[m.Name()] = m
+
+	return nil
+}
+
+// forward receives Event by all Brokers and sends them over one single channel to the Router
+func (r *Router) forward(in <-chan *Event) {
+	for {
+		select {
+		case msg := <-in:
```
Comment (fortytw2) [inline main.go]: if kept, would need a check for the channel being closed

File: slack.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"github.com/go-kit/kit/log"
+	"github.com/nlopes/slack"
+)
+
+// SlackBroker implements the Broker interface and acts as a Slack Bot
+type SlackBroker struct {
+	Config Config
```
Comment (fortytw2) [inline slack.go]: probably best to not propogate the global configuration down this far, but to provide slack-specific settings that are set on initialization

`NewSlackBroker(l log.Logger, c *slack.Client) (Broker, error)`

File: slack.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Send takes an Event and sends it via this Broker
+func (b *SlackBroker) Send(event *Event) {
+	_, _, err := b.client.PostMessage(event.Channel, event.Body, slack.PostMessageParameters{
+		AsUser: true,
+	})
+	if err != nil {
+		b.Logger.Log(
+			"level", "warm",
```
Comment (fortytw2) [inline slack.go]: if you want leveled logging, might be good to use and pass around a https://godoc.org/github.com/go-kit/kit/log/levels

File: event.go (hunk 1 above)
Comment (metalmatze) [inline event.go]: No, right now they are used to send via the broker too.
`broker.send(&event)`. But really I'm really not sure about passing `Event{}` everywhere in general.

Comment (fortytw2) [inline event.go]: I think using different structs for Send / Receive might be beneficial, especially as they start to diverge down the line...

File: slack.go (hunk 7 above)
Comment (dominikschulz) [inline slack.go]: That's are rather intrusive interface. I'd prefer to just write out the log levels if we need them.

Comment (fortytw2) [inline slack.go]: Ideally, just pass a NoOp logger when you don't need logs :)

Comment (metalmatze) [inline slack.go]: I totally understand the NoOp logger concept. But how do you change the verbosity if you want to debug? Add `--debug` to the command? I think log levels are there to enable filtering afterwards. If someone would now want to debug this binary they'd have to build a new binary.

**Pull request merged**